### PR TITLE
Rails API and React Finals

### PR DESCRIPTION
**Pull request by AlphaNtihinduka**
# Project requirements
```
- We are still having the issues with completing the 100% project. 
So far, when creating the new car and reservation, you need to manually provide the user_id.
- We are also still working on hiding navigation bar before user login. Thank you.
````

## Basics

- You should follow the layout of [the provided design](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign). You should change only the titles, descriptions, and photos - in order to create a website about something other than motorcycles.
- Select a theme for your website - is it going to be a website for booking doctor appointments, booking online classes, or something else?

### Core features - for teams of each size

- The user logs in to the website, only by typing the username (a proper authenticated login is a requirement if your group is made of 5 people).

- In the navigation panel, the user can see links to:

  - Motorcycles/doctors/classes/items that you selected as a theme.
  - "Reserve" form.
  - "My reservations".
  - "Add motorcycle/doctor/class/item that you selected as a theme" (the link is visible to everybody).
  - "Delete motorcycle/doctor/class/item that you selected as a theme" (the link is visible to everybody).
 - On the main page, the user can see a [list of motorcycles/doctors/classes/items that you selected as a theme](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005577).

List

- When the user selects a specific item, they can see the [details page](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005579) with its full description (skip the "Rotate image" button).

- In the details page, the user can click the "Reserve" button (in the design you can see the "Configure" button - please replace it with the "Reserve" button).
Details

- When the user clicks the "Add item" link in the navigation panel they can see a form for adding a new item.

- Make the app responsive, creating both mobile and desktop versions.

### If your team has 3-4 members these are the required features; otherwise, consider them as optional features.

- When the user clicks the "Delete item" link in the navigation panel they can see a list of all items with title and "Delete" button.
- When the user clicks the "Delete" button, the selected item is marked as removed and does not show on the main list anymore.
- To reserve an appointment, the user has to select a date and city (username and selected item are autofilled).

  - Use the design based on the ["Book a vespa test-ride"](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005583) and add all necessary inputs.
  - The user can also access the "Reserve" page from the navigation panel. In that case only username is autofilled.

Booking

- When the user clicks the "My reservations" link in the navigation panel they can see a list of their reservations (with information about item name, date and city).
- Add full documentation for your API.

## Technical set up

- Set up the repository/repositories on GitHub and use Gitflow.
- You should use Postgres as your database
- Use Rails to create backend API.
- Use React & Redux to create frontend UI.
- You can choose if you want to set up your project as two separate apps or as one.
- The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  mount Rswag::Ui::Engine => '/api-docs'
+  mount Rswag::Api::Engine => '/api-docs'
+ devise_for :users
+
+  post 'api/v1/auth/login', to: 'authentication#login'
+  namespace :api do
+    namespace :v1 do
+       resources :users
+       resources :cars do
+       resources :reservations
```
Comment (vichuge) [inline config/routes.rb:11]: - I understand you're not using all endpoints for each controller, I kindly invite you to use the `:only` command, to declare what are the endpoints you will use for each controller. 🐔

File: app/controllers/api/v1/reservations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
+class Api::V1::ReservationsController < ApplicationController
+  def index
```
Comment (vichuge) [inline app/controllers/api/v1/reservations_controller.rb:2]: - It looks like this controller could be used for anybody without some kind of restriction, I invite you to add the `before_action` command on `reservations_controllers.rb` to be used only for logged users. 🐱

File: spec/model/car_spec.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
```
Comment (vichuge) [inline spec/model/car_spec.rb:1]: - It looks like there are some issues to fix on tests in general, I'll add here the failed test, please, I invite you to fix all of this. 🐧 

```
30 examples, 12 failures, 1 error occurred outside of examples

Failed examples:

rspec ./spec/requests/api/v1/cars_spec.rb:19 # api/v1/cars /api/v1/cars post successful returns a 200 response
rspec ./spec/requests/api/v1/cars_spec.rb:34 # api/v1/cars /api/v1/cars/new get successful returns a 200 response
rspec ./spec/requests/api/v1/cars_spec.rb:52 # api/v1/cars /api/v1/cars/{id}/edit get successful returns a 200 response
rspec ./spec/requests/api/v1/cars_spec.rb:73 # api/v1/cars /api/v1/cars/{id} get name found returns a 200 response
rspec ./spec/requests/api/v1/cars_spec.rb:88 # api/v1/cars /api/v1/cars/{id} get Car not found returns a 404 response
rspec ./spec/requests/api/v1/reservations_spec.rb:19 # api/v1/reservations /api/v1/cars/{car_id}/reservations get successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:42 # api/v1/reservations /api/v1/cars/{car_id}/reservations post successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:62 # api/v1/reservations /api/v1/cars/{car_id}/reservations/new get successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:83 # api/v1/reservations /api/v1/cars/{car_id}/reservations/{id}/edit get successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:121 # api/v1/reservations /api/v1/cars/{car_id}/reservations/{id} patch successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:137 # api/v1/reservations /api/v1/cars/{car_id}/reservations/{id} put successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:153 # api/v1/reservations /api/v1/cars/{car_id}/reservations/{id} delete successful returns a 200 response
```

File: spec/model/reservation_spec.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+require 'rails_helper'
```
Comment (vichuge) [inline spec/model/reservation_spec.rb:1]: - It looks like there are some issues to fix on tests in general, I'll add here the failed test, please, I invite you to fix all of this. 🐧 

```
30 examples, 12 failures, 1 error occurred outside of examples

Failed examples:

rspec ./spec/requests/api/v1/cars_spec.rb:19 # api/v1/cars /api/v1/cars post successful returns a 200 response
rspec ./spec/requests/api/v1/cars_spec.rb:34 # api/v1/cars /api/v1/cars/new get successful returns a 200 response
rspec ./spec/requests/api/v1/cars_spec.rb:52 # api/v1/cars /api/v1/cars/{id}/edit get successful returns a 200 response
rspec ./spec/requests/api/v1/cars_spec.rb:73 # api/v1/cars /api/v1/cars/{id} get name found returns a 200 response
rspec ./spec/requests/api/v1/cars_spec.rb:88 # api/v1/cars /api/v1/cars/{id} get Car not found returns a 404 response
rspec ./spec/requests/api/v1/reservations_spec.rb:19 # api/v1/reservations /api/v1/cars/{car_id}/reservations get successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:42 # api/v1/reservations /api/v1/cars/{car_id}/reservations post successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:62 # api/v1/reservations /api/v1/cars/{car_id}/reservations/new get successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:83 # api/v1/reservations /api/v1/cars/{car_id}/reservations/{id}/edit get successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:121 # api/v1/reservations /api/v1/cars/{car_id}/reservations/{id} patch successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:137 # api/v1/reservations /api/v1/cars/{car_id}/reservations/{id} put successful returns a 200 response
rspec ./spec/requests/api/v1/reservations_spec.rb:153 # api/v1/reservations /api/v1/cars/{car_id}/reservations/{id} delete successful returns a 200 response
```

**Review by vichuge — COMMENTED** [review]
Hi dear student, 

Good job so far!
You still need to work on some issues to prepare your project for the final evaluation, but you are almost there!

mario_ladders


## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you take them into account as they can make your code better._

## ⚠️ ⚠️ ⚠️  I'm afraid I wasn't able to find the Front-end URL repo. Please, add it for the next code reviewer. I checked the project readme file and PR description but It wasn't there. Thanks for your work!

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Comment by 41hulk** [comment]
On this project, we used one repo for the backend and the front end which means there's no Front-end repo URL. You can find the Front-end part in `./app/javascript`

**Pull request merged**